sel/src/Sel/PublicKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,494 @@
+
+-- | > show secretKey == "[REDACTED]"
+--
+-- @since 0.0.1.0
+instance Show SecretKey where
+  show _ = "[REDACTED]"
+
+-- | A public key of size 'cryptoBoxPublicKeyBytes'.
+--
+-- @since 0.0.1.0
+newtype PublicKey = PublicKey (ForeignPtr CUChar)
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+      -- > display secretKey == "[REDACTED]"
+    )
+    via (OpaqueInstance "[REDACTED]" PublicKey)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq PublicKey where
+  (PublicKey hk1) == (PublicKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoBoxPublicKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord PublicKey where
+  compare (PublicKey hk1) (PublicKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoBoxPublicKeyBytes
+
+-- | > show secretKey == "[REDACTED]"
+--
+-- @since 0.0.1.0
+instance Show PublicKey where
+  show _ = "[REDACTED]"
+
+data KeyPairGenerationException = KeyPairGenerationException
+  deriving stock (Eq, Ord, Show)
+  deriving anyclass (Exception)
+
+-- | Generate a new random secret key.
+--
+-- @since 0.0.1.0
+newKeyPair :: IO (SecretKey, PublicKey)
+newKeyPair = newKeyPairWith $ \secretKeyPtr publicKeyPtr -> do
+  result <- cryptoBoxKeyPair secretKeyPtr publicKeyPtr
+  when (result /= 0) $ throw KeyPairGenerationException
+
+-- | Create a 'SecretKey' from a binary 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- @since 0.0.1.0
+secretKeyPairFromByteStrings
+  :: StrictByteString
+  -- ^ Secret key
+  -> StrictByteString
+  -- @ Public keyu
+  -> Maybe (SecretKey, PublicKey)
+secretKeyPairFromByteStrings secretByteString publicByteString =
+  if BS.length secretByteString < fromIntegral cryptoBoxSecretKeyBytes
+    || BS.length publicByteString < fromIntegral cryptoBoxPublicKeyBytes
+    then Nothing
+    else unsafeDupablePerformIO $
+      BS.unsafeUseAsCString secretByteString $ \outsideSecretKeyPtr ->
+        BS.unsafeUseAsCString publicByteString $ \outsidePublicKeyPtr ->
+          fmap Just $
+            newKeyPairWith $ \secretKeyPtr publicKeyPtr -> do
+              Foreign.copyArray
+                outsideSecretKeyPtr
+                (Foreign.castPtr @CUChar @CChar secretKeyPtr)
+                (fromIntegral cryptoBoxSecretKeyBytes)
+
+              Foreign.copyArray
+                outsidePublicKeyPtr
+                (Foreign.castPtr @CUChar @CChar publicKeyPtr)
+                (fromIntegral cryptoBoxPublicKeyBytes)
+
+-- | Prepare memory for a 'SecretKey' and 'PublicKey' pair,
+-- and use the provided action to fill it.
+--
+-- Memory is allocated with 'LibSodium.Bindings.SecureMemory.sodiumMalloc' (see the note attached there).
+-- Finalizer is run when the key is goes out of scope, but 'freeSecretKey' can be used to release early.
+--
+-- @since 0.0.1.0
+newKeyPairWith
+  :: ( Ptr CUChar
+       -- \^ Secret Key pointer
+       -> Ptr CUChar
+       -- \^ Public Key pointer
+       -> IO ()
```

Comment:
Since libsodium always puts the public key first, it's probably best to follow that convention for the callback as well.